tests/AssertAnnotationTest.php

```diff
@@ -437,6 +437,87 @@ function assertIntOrFoo($b) : void {
 
                     if (!is_int($a)) $a->bar();',
             ],
+            'unionOfStringTypes' => [
+                '<?php
+                /** @psalm-assert "self"|class-string $b */
+                function a(string $b): void {
+                    if ("" === $b) { throw new \Exception("");};
+                }
+
+                /** @psalm-assert-if-true "self"|class-string $b */
+                function atrue(string $b): bool {
+                    return $b === "";
+                }
+
+                /** @psalm-assert-if-false "self"|class-string $b */
+                function afalse(string $b): bool {
+                    return $b === "";
+                }
+
+                /** @param "self"|class-string $arg */
+                function check(string $arg): void {}
+
+                /** @var string $var */;
+                a($var);
+
+                /** @var string $varTrue */;
+                if (!atrue($varTrue)) {
+                    exit;
+                }
+
+                /** @var string $varFalse */;
+                if (afalse($varFalse)) {
+                    exit;
+                }
+
+                check($var);
+                check($varTrue);
```

Comment:
This one is not working as you think it is. I believe `$varTrue` is just `class-string` instead of `"self"|class-string` but check() accepts `class-string` as a valid value for `"self"|class-string` so it's not giving you errors. Your assertion test is more trustworthy

Reply:
That's exactly what I tried to test — trace shows `class-string|"self"` for `$var`, but just `class-string` for -if* assertions.